apps/teca_regional_moisture_flux.cpp

```diff
@@ -0,0 +1,612 @@
+    }
+
+    if (!opt_vals["y_axis_variable"].defaulted())
+    {
+        cf_reader->set_y_axis_variable(opt_vals["y_axis_variable"].as<string>());
+        mcf_reader->set_y_axis_variable(opt_vals["y_axis_variable"].as<string>());
+    }
+
+    // add the valid value mask stage
+    norm_coords->set_input_connection(head->get_output_port());
+    vv_mask->set_input_connection(norm_coords->get_output_port());
+    unpack->set_input_connection(vv_mask->get_output_port());
+    head = unpack;
+
+    if (opt_vals.count("compute_ivt"))
+    {
+        if (mpi_man.get_comm_rank() == 0)
+        {
+            TECA_STATUS("Computing IVT")
+        }
+
+        std::string z_var = "plev";
+        if (!opt_vals["z_axis_variable"].defaulted())
+            z_var = opt_vals["z_axis_variable"].as<string>();
+
+        cf_reader->set_z_axis_variable(z_var);
+        mcf_reader->set_z_axis_variable(z_var);
+
+        ivt_int->set_input_connection(head->get_output_port());
+        ivt_int->set_wind_u_variable(opt_vals["wind_u"].as<string>());
+        ivt_int->set_wind_v_variable(opt_vals["wind_v"].as<string>());
+        ivt_int->set_specific_humidity_variable(opt_vals["specific_humidity"].as<string>());
+        ivt_int->set_ivt_u_variable(opt_vals["ivt_u"].as<string>());
+        ivt_int->set_ivt_v_variable(opt_vals["ivt_v"].as<string>());
+
+        head = ivt_int;
+    }
+/*
+
+    // configure the reader
+    if (!opt_vals.count("mask_regex"))
+    {
+        if (mpi_man.get_comm_rank() == 0)
+        {
+            TECA_FATAL_ERROR("The region mask file was not specified. See --mask_regex")
+        }
+        return -1;
+    }
+
+    if (!opt_vals.count("mask"))
+    {
+        if (mpi_man.get_comm_rank() == 0)
+        {
+            TECA_FATAL_ERROR("The region mask variable was not specified. See --mask")
+        }
+        return -1;
+    }
+
+
+
+    mask_reader->set_files_regex(opt_vals["mask_regex"].as<string>());
+
+    mask_coords->set_input_connection(mask_reader->get_output_port());
+
+    mask_vvm->set_input_connection(mask_coords->get_output_port());
+*/
+    // region mask goes in a separate branch of the pipeline so that it can be cached
+    teca_metadata md = head->update_metadata();
+
+    mask_mesh->set_spatial_bounds(md, false);
+    mask_mesh->set_spatial_extents(md, false);
+    mask_mesh->set_x_axis_variable(md);
+    mask_mesh->set_y_axis_variable(md);
+    //mask_mesh->set_z_axis_variable(md);
+    mask_mesh->set_t_axis_variable(md);
+    mask_mesh->set_t_axis(md);
+
+    // shape file mask
+    std::string region_name = opt_vals["region_name"].as<string>();
+
+    mask->set_input_connection(mask_mesh->get_output_port());
+    mask->set_shape_file(opt_vals["shape_file"].as<string>());
+    mask->set_mask_variable(region_name + "_mask");
+
+    mask_cache->set_input_connection(mask->get_output_port());
+
+    mask_regrid->set_input_connection(0, head->get_output_port());
```

Comment:
At some point (in another PR), it might be worth separating out the functionality being used here in  `teca_cartesian_mask_regrid`: maybe a new algorithm like `teca_pipeline_join`.  The only reason would be to make it more clear to a power user how to combine the dataset cache.  On a technical side, we also might find some use cases where we want to have multiple branches of a pipeline upstream of an algorithm (maybe a cf_reader, a table_reader, and a dataset cache).  The `teca_pipeline_join` algorithm could be configured to accept a flexible number of input connections.

Reply:
Yes, assuming the meshes have the same spatial coordinate system, then we could then join any number of pipeline branches. Great idea. 

Also, we really need more documentation of what's already implemented and how to use.

Reply:
Captured in issue #743